tools/plugin/modules/noise_suppression_interface.cpp

```diff
@@ -0,0 +1,164 @@
+		int iter;
+	};
+
+	int ov_ns_init(ns_handle *handle) {
+		struct ns_data *nd;
+		ov::OutputVector inputs, outputs;
+		ov::Core core;
+		size_t state_size = 0;
+		const char* model_name = std::getenv("NOISE_SUPPRESSION_MODEL_NAME");
+		int i;
+
+		nd = new ns_data();
+		if(!nd)
+			return -ENOMEM;
+
+		*handle = nd;
+		nd->model = core.read_model(model_name);
+		inputs = nd->model->inputs();
+		outputs = nd->model->outputs();
+
+		/* get state name pairs */
+		for (size_t i = 0; i < inputs.size(); i++) {
+			std::string inp_state_name = inputs[i].get_any_name();
+			if (inp_state_name.find("inp_state_") == std::string::npos)
+			    continue;
+
+			std::string out_state_name(inp_state_name);
+			out_state_name.replace(0, 3, "out");
+
+			/* find corresponding output state */
+			if (outputs.end() == std::find_if(outputs.begin(), outputs.end(), [&out_state_name](const ov::Output<ov::Node>& output) {
+				return output.get_any_name() == out_state_name;
+			}))
+				return -EINVAL;
+
+			nd->state_names.emplace_back(inp_state_name, out_state_name);
+
+			ov::Shape shape = inputs[i].get_shape();
+			size_t tensor_size = std::accumulate(shape.begin(), shape.end(), 1,
+							     std::multiplies<size_t>());
+
+			state_size += tensor_size;
+		}
+
+		if (!state_size)
+			return -EINVAL;
+
+		/*
+		 * compile the model for the CPU and save the infer_request objects for each
+		 * channel separately
+		 */
+		ov::CompiledModel compiled_model = core.compile_model(nd->model, "CPU", {});
+		for (i = 0; i < NS_MAX_SOURCE_CHANNELS; i++)
+			nd->infer_request[i] = compiled_model.create_infer_request();
+
+		nd->inp_shape = nd->model->input("input").get_shape();
+
+		return 0;
+	}
+
+	void ov_ns_free(ns_handle *handle) {
+		struct ns_data *nd = (struct ns_data *)handle;
+
+		delete nd;
+	}
+
+	int ov_ns_process(ns_handle handle,
+			  struct input_stream_buffer *input_buffers, int num_input_buffers,
+			  struct output_stream_buffer *output_buffers, int num_output_buffers)
+	{
+		struct audio_stream *source = (struct audio_stream *)input_buffers[0].data;
+		struct audio_stream *sink = (struct audio_stream *)output_buffers[0].data;
+		struct ns_data *nd = (struct ns_data *)handle;
+		std::vector<float> inp_wave_fp32, out_wave_fp32;
+		/* only 16-bit supported for now */
+		int16_t *input_data = (int16_t *)audio_stream_get_rptr(source);
+		int16_t *output_data = (int16_t *)audio_stream_get_wptr(sink);
+		uint32_t frame_count = input_buffers[0].size;
+		int i, j, ch;
+
+		float scale = 1.0f / std::numeric_limits<int16_t>::max();
+
+		/*
+		 * The noise suppression model only supports mono, so process each channel
+		 * separately
+		 */
+		inp_wave_fp32.resize(frame_count, 0);
+		out_wave_fp32.resize(frame_count, 0);
+
+		for (ch = 0; ch < NS_MAX_SOURCE_CHANNELS; ch++) {
+			/* split each channel samples and convert to floating point */
+			for (i = ch, j = 0; j < frame_count; i+=2,j++)
+				inp_wave_fp32[j] = (float)input_data[i] * scale;
```

Comment:
Is there no need to check input and output buffer for wrap? I can't see a module interface struct populate, but this is still legacy process type and not sink/source API, right?

Reply:
yes, legacy process. Do you think sink/source makes more sense with this?

Reply:
As long as legacy is not deprecated should be no need. The sink/source API allows resources savings in DSP but it's not as critical on the host side. Also we don't yet have those savings implemented so currently there's no difference. The no-wrap check linear buffers will be a feature in sink/source API.

I was asking if it's ensured in plugin that buffers can be operated as linear even in legacy. It might work so too in FW but I'm not sure it's ensured. The components have still the samples/frame before wrap functions in use.

Reply:
@singalsu I've added the checks for input/output wrap now.